source/common/common/inline_map.h

```diff
@@ -0,0 +1,495 @@
+      if (inlineEntryValid(*entry_id)) {
+        return {inline_entries_[*entry_id], false};
+      }
+
+      resetInlineMapEntry(*entry_id, std::forward<SetValue>(value));
+      return {inline_entries_[*entry_id], true};
+    } else {
+      // This key is not registered as inline key and try to insert the value to the normal map.
+      auto result =
+          dynamic_entries_.emplace(std::forward<SetKey>(key), std::forward<SetValue>(value));
+      return {result.first->second, result.second};
+    }
+  }
+
+  /**
+   * Set the entry by the given handle.
+   * @param handle the handle to set.
+   * @param value the value to set.
+   * @return pair consisting of a reference to the element being set, or the already-existing
+   * element if no setting happened, and a bool denoting whether the setting took place (true if
+   * setting happened, false if it did not).
+   */
+  template <class SetValue> std::pair<ValueRef, bool> set(Handle handle, SetValue&& value) {
+    // If the entry is already valid, insert will fail and return false.
+    if (inlineEntryValid(handle.inlineId())) {
+      return {inline_entries_[handle.inlineId()], false};
+    }
+
+    resetInlineMapEntry(handle.inlineId(), std::forward<SetValue>(value));
+    return {inline_entries_[handle.inlineId()], true};
+  }
+
+  /**
+   * Erase the entry by the given key. If the key is not found, do nothing.
+   * Heterogeneous lookup is supported here.
+   * @param key the key to erase.
+   * @return the number of elements erased.
+   */
+  template <class GetKey> uint64_t erase(const GetKey& key) {
+    if (absl::optional<uint64_t> entry_id = inlineLookup(key); entry_id.has_value()) {
+      return clearInlineMapEntry(*entry_id);
+    } else {
+      return dynamic_entries_.erase(key);
+    }
+  }
+
+  /**
+   * Erase the entry by the given handle. If the handle is not valid, do nothing.
+   * @param handle the handle to erase.
+   * @return the number of elements erased.
+   */
+  uint64_t erase(Handle handle) { return clearInlineMapEntry(handle.inlineId()); }
+
+  /**
+   * Iterate all elements in the map.
+   * @param callback the callback function to handle each element.
+   */
+  void iterate(std::function<bool(const Key&, const Value&)> callback) const {
+    for (const auto& entry : dynamic_entries_) {
+      if (!callback(entry.first, entry.second)) {
+        return;
+      }
+    }
+
+    const uint64_t inline_keys_num = descriptor_.inlineKeysNum();
+    const auto& inline_keys = descriptor_.inlineKeys();
+    for (uint64_t i = 0; i < inline_keys_num; ++i) {
+      if (!inlineEntryValid(i)) {
+        continue;
+      }
+
+      if (!callback(inline_keys[i], inline_entries_[i])) {
+        return;
+      }
+    }
+  }
+
+  /**
+   * Clear all elements in the map.
+   */
+  void clear() {
+    // Destruct all inline entries.
+    for (uint64_t i = 0; i < descriptor_.inlineKeysNum(); ++i) {
```

Comment:
hoist the call to inlineKeysNum out of the loop like you do on line 342 (slight performance improvement)